Sparse CSR CUDA: Add torch.baddbmm and torch.bmm

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #68711

This PR adds possibility to multiply a single CSR matrix by a batch of dense matrices.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ngimel

Differential Revision: [D33773319](https://our.internmc.facebook.com/intern/diff/D33773319)

IvanYashchuk committed:
Sparse CSR CUDA: Add torch.baddbmm and torch.bmm

This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiple a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -585,6 +585,71 @@ def test_csr_matvec(self, device, dtype):
+                a = self.genSparseCSRTensor((m, k), nnz, dtype=dtype, device=device, index_dtype=index_dtype)
+
+                # a_batched is a regular CSR tensor but with a batch dimension in the shape
+                a_batched = torch._sparse_csr_tensor_unsafe(
```
So this is effectively a "hack" to circumvent the shape check that baddmm requires, because in the underlying C++ code you constraint yourself to the first batch entry of the CSR matrix? I guess if we allowed broadcasting in the batch dimensions this wouldn't be necessary and batch_size could be 1.

IvanYashchuk replied:
Right, it's a hack to circumvent the shape check. `matmul` allows broadcasting batch dimension, but `bmm` doesn't.
```
In [1]: import torch
In [2]: a = torch.randn(3, 3)
In [3]: b = torch.randn(2, 3, 3)
In [4]: a @ b
Out[4]:
tensor([[[-2.3947,  4.1253,  1.6287],
         [-0.8864,  0.6599, -0.7547],
         [ 5.3741, -4.8693,  2.2964]],

        [[ 4.5006,  0.3885, -2.6066],
         [ 0.4729,  0.0400, -0.4279],
         [-5.1257, -0.6420,  3.7720]]])
In [5]: a.bmm(b)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-5-bd5c7f2de399> in <module>
----> 1 a.bmm(b)

RuntimeError: batch1 must be a 3D tensor
In [6]: a.unsqueeze(0).bmm(b)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-5dac428fe979> in <module>
----> 1 a.unsqueeze(0).bmm(b)

RuntimeError: Expected size for first two dimensions of batch2 tensor to be: [1, 3] but got: [2, 3].
```
I need to do something so that this hack is not required.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33773319).

cpuhrsch commented:
@pytorchbot ciflow rerun -l ciflow/all

cpuhrsch commented:
@IvanYashchuk - could you rebase please?

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiply a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D33773319](https://our.internmc.facebook.com/intern/diff/D33773319)

[ghstack-poisoned]

IvanYashchuk commented:
@cpuhrsch, rebased.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33773319).

cpuhrsch commented:
@IvanYashchuk - unfortunately the errors seem legit. [libtorch-linux-xenial-cuda10.2-py3.7-gcc7 ](https://github.com/pytorch/pytorch/runs/4942866873?check_suite_focus=true) complains that

```
/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cpp
2022-01-25T20:55:23.6359632Z In file included from [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDAContext.h:17:0[m[K,
2022-01-25T20:55:23.6360361Z                  from [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:1[m[K:
2022-01-25T20:55:23.6392841Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:[m[K In constructor ‘[01m[Kat::cuda::sparse::CuSparseSpMatCsrDescriptor::CuSparseSpMatCsrDescriptor(const at::Tensor&)[m[K’:
2022-01-25T20:55:23.6437569Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:172:30:[m[K [01;31m[Kerror: [m[K‘[01m[KcusparseCsrSetStridedBatch[m[K’ was not declared in this scope
2022-01-25T20:55:23.6438281Z        TORCH_CUDASPARSE_CHECK([01;31m[Kc[m[KusparseCsrSetStridedBatch(
2022-01-25T20:55:23.6438585Z                               [01;31m[K^[m[K
2022-01-25T20:55:23.6439094Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/Exceptions.h:60:30:[m[K [01;36m[Knote: [m[Kin definition of macro ‘[01m[KTORCH_CUDASPARSE_CHECK[m[K’
2022-01-25T20:55:23.6439529Z      cusparseStatus_t __err = [01;36m[KEXPR[m[K;                              \
2022-01-25T20:55:23.6439803Z                               [01;36m[K^~~~[m[K
2022-01-25T20:55:23.6440337Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:172:30:[m[K [01;36m[Knote: [m[Ksuggested alternative: ‘[01m[KcusparseDnMatSetStridedBatch[m[K’
2022-01-25T20:55:23.6440825Z        TORCH_CUDASPARSE_CHECK([01;36m[Kc[m[KusparseCsrSetStridedBatch(
2022-01-25T20:55:23.6441124Z                               [01;36m[K^[m[K
2022-01-25T20:55:23.6441549Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/Exceptions.h:60:30:[m[K [01;36m[Knote: [m[Kin definition of macro ‘[01m[KTORCH_CUDASPARSE_CHECK[m[K’
2022-01-25T20:55:23.6441973Z      cusparseStatus_t __err = [01;36m[KEXPR[m[K;                              \
2022-01-25T20:55:23.6442264Z                               [01;36m[K^~~~[m[K
2022-01-25T20:55:23.6442952Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:181:11:[m[K [01;31m[Kerror: [m[K‘[01m[KcusparseCsrSetStridedBatch[m[K’ was not declared in this scope
2022-01-25T20:55:23.6443464Z            [01;31m[Kc[m[KusparseCsrSetStridedBatch(raw_descriptor, batch_count, 0, 0));
2022-01-25T20:55:23.6443756Z            [01;31m[K^[m[K
2022-01-25T20:55:23.6444172Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/Exceptions.h:60:30:[m[K [01;36m[Knote: [m[Kin definition of macro ‘[01m[KTORCH_CUDASPARSE_CHECK[m[K’
2022-01-25T20:55:23.6444591Z      cusparseStatus_t __err = [01;36m[KEXPR[m[K;                              \
2022-01-25T20:55:23.6444875Z                               [01;36m[K^~~~[m[K
2022-01-25T20:55:23.6445375Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDASparseDescriptors.cpp:181:11:[m[K [01;36m[Knote: [m[Ksuggested alternative: ‘[01m[KcusparseDnMatSetStridedBatch[m[K’
2022-01-25T20:55:23.6446163Z            [01;36m[Kc[m[KusparseCsrSetStridedBatch(raw_descriptor, batch_count, 0, 0));
2022-01-25T20:55:23.6446452Z            [01;36m[K^[m[K
2022-01-25T20:55:23.6446888Z [01m[K/var/lib/jenkins/workspace/aten/src/ATen/cuda/Exceptions.h:60:30:[m[K [01;36m[Knote: [m[Kin definition of macro ‘[01m[KTORCH_CUDASPARSE_CHECK[m[K’
2022-01-25T20:55:23.6447294Z      cusparseStatus_t __err = [01;36m[KEXPR[m[K;                              \
2022-01-25T20:55:23.6447576Z                               [01;36m[K^~~~[m[K
```

This has to be pulled from the raw logs (hence the escape symbols)

IvanYashchuk committed:
Update on "Sparse CSR CUDA: Add torch.baddbmm and torch.bmm"


This PR adds possibility to multiply a single CSR matrix by a batch of dense matrices.

cc nikitaved pearu cpuhrsch IvanYashchuk ngimel

Differential Revision: [D33773319](https://our.internmc.facebook.com/intern/diff/D33773319)

[ghstack-poisoned]

IvanYashchuk commented:
@cpuhrsch, thanks for pointing out the CUDA 10.2 failure! For some reason, I had in mind that the support for versions older than 11 was dropped and forgot to add the guards. Now they're there.

cpuhrsch commented:
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33773319).
